Mathlib/Topology/UrysohnsLemma.lean

```diff
@@ -465,3 +468,253 @@ theorem exists_continuous_nonneg_pos [RegularSpace X] [LocallyCompactSpace X] (x
+            rw [add_comm 1 m]
+            exact Nat.lt_add_one_iff
+          rw [hxlem]
+          ring
+          all_goals
+            simp only [Finset.mem_filter, Finset.mem_univ, true_and, not_le]
+            rw [Fin.lt_def]
+            simp only [Fin.val_natCast]
+            exact lt_add_one m
+      intro x hx
+      have huniv : {j : Fin (Nat.succ n) | j ≤ ⟨n, (lt_add_one n)⟩ }.toFinset = Finset.univ := by
+        ext j
+        simp only [Nat.zero_eq, mem_setOf_eq, toFinset_setOf, Finset.mem_filter, Finset.mem_univ,
+          true_and, iff_true]
+        apply Fin.mk_le_of_le_val
+        simp only
+        exact Nat.lt_succ_iff.mp j.isLt
+      rw [← huniv]
+      have heqfun (x : X) :
+          (∑ j in { j : Fin (Nat.succ n) | j ≤ ⟨n, (lt_add_one n)⟩ }.toFinset, f j) x
+          = (1 - (∏ j in { j : Fin (Nat.succ n) | j ≤ ⟨n, (lt_add_one n)⟩ }.toFinset, (1 - g j)))
+          x := by
+        apply Function.funext_iff.mp
+        ext z
+        exact congrFun (congrArg DFunLike.coe (hsumf n (lt_add_one n))) z
+      simp only [Nat.zero_eq, mem_setOf_eq, toFinset_setOf, ContinuousMap.coe_sum, Finset.sum_apply,
+        ContinuousMap.sub_apply, ContinuousMap.one_apply, ContinuousMap.coe_prod,
+        ContinuousMap.coe_sub, ContinuousMap.coe_one, Finset.prod_apply, Pi.sub_apply,
+        Pi.one_apply] at heqfun
+      simp only [Nat.zero_eq, mem_setOf_eq, toFinset_setOf, Finset.sum_apply, Pi.one_apply]
+      rw [heqfun]
+      simp only [sub_eq_self]
+      obtain ⟨xj, hxj⟩ := exists_exists_eq_and.mp (hι.2 hx)
+      simp only [mem_iUnion, exists_prop] at hxj
+      rw [hW] at hxj
+      simp at hxj
+      rw [dif_pos (hι.1 xj hxj.1)] at hxj
+      obtain ⟨i, hi⟩ := (Classical.choose_spec (htW xj (hι.1 xj hxj.1))).2.2.2
+      have hxHi : x ∈ H i := by
+        rw [hH]
+        simp only [mem_iUnion]
+        use ⟨xj, hxj.1⟩
+        rw [hWx]
+        simp only [dite_eq_ite]
+        rw [hW]
+        simp only
+        rw [dif_pos (hι.1 xj hxj.1)]
+        apply Set.mem_of_mem_of_subset _ subset_closure
+        simp only [mem_ite_empty_right]
+        exact And.intro hi (Set.mem_of_mem_of_subset hxj.2 subset_closure)
+      simp at huniv
+      rw [huniv]
+      apply Finset.prod_eq_zero (Finset.mem_univ i)
+      rw [hg]
+      simp only [mem_Icc]
+      rw [(Classical.choose_spec (exists_tsupport_one_of_isOpen_isClosed
+        (hs i) (IsClosedH i) (IsHSubS i))).2.1 hxHi]
+      simp only [Pi.one_apply, sub_self]
+    · intro i x
+      rw [hf]
+      simp only [mem_setOf_eq, toFinset_setOf, ContinuousMap.mul_apply, ContinuousMap.coe_prod,
+        ContinuousMap.coe_sub, ContinuousMap.coe_one, Finset.prod_apply]
+      apply unitInterval.mul_mem
+      · set g' : Fin (n.succ) → ℝ := fun c => (1 - (g c)) x
+        have hg' : ∀ (c : Fin (n.succ)), ((1 : X → ℝ) - (g c)) x = g' c := by
+          intro c
+          rfl
+        simp_rw [hg']
+        refine prod_mem (S := unitInterval.unitIntervalSubmonoid) ?_
```

Comment:
Can you extract this to a standalone `unitInterval.prod_mem` theorem in the other file?

Reply:
sure, done.